src/osd/ECCache.hpp

```diff
@@ -0,0 +1,273 @@
+    bool operator < (const class ecc_object& r) const
+    {
+      return key < r.key;
+    }
+
+    bool operator == (const class ecc_object& r) const
+    {
+      return key == r.key;
+    }
+
+    ecc_object(std::string k = "", bool eflag = false, extents ein = {})
+    {
+      evictable = eflag;
+      key.assign(k);
+      e = ein;
+    }
+
+    ~ecc_object(){}
+
+    std::string get_key()
+    {
+      return key;
+    }
+
+    void evict()
+    {
+      evictable = true;
+    }
+
+    bool is_evictable()
+    {
+      return evictable;
+    }
+
+    int put_extent(unsigned o, bufferlist d)
+    {
+      unsigned l(d.length());
+      unsigned nso(o), neo(o+l-1);
+      bufferlist nd;
+
+      extents::iterator low = e.upper_bound(o);
+      extents::iterator up = e.upper_bound(o+l-1);
+
+      if((!e.empty()) && low != e.begin()){
+	--low;
+	nso = (low->end() >= nso) ? low->start() : nso;
+	nd.substr_of(low->data(), 0, o-nso);
+      }
+      bufferlist tmp;
+      tmp.substr_of(d, 0, l);
+      nd.append(tmp);
+      if((!e.empty()) && up != e.begin()){
+	--up;
+	neo = (up->end() > neo) ? up->end() : neo;
+	bufferlist tmp1;
+	tmp1.substr_of(up->data(), (up->length() - (neo - (o+l-1))), (neo - (o+l-1)));
+	nd.append(tmp1);
+	++up;
+      }
+      e.erase(low, up);
+      ecc_extent tmp_extent(nso, nd);
+      e.insert(tmp_extent);
+      return (l);
+    }
+
+    int get_extent( unsigned o, unsigned l, extents *out)
+    {
+      bufferlist tmp;
+      unsigned t, r = l;
+
+      for(extents::iterator i = e.begin(); i != e.end(); i++){
+	if(i->end() < o) continue;
+	if(i->start() < o){
+	  t = ((i->length()) - (o-(i->start())));
+	  t = t > r ? r : t;
+	  tmp.substr_of(i->data(), o-(i->start()), t);
+	  ecc_extent tmp_extent(o, tmp);
+	  out->insert(tmp_extent);
+	  r -=t;
+	  continue;
+	}
+	if(i->end() > (o+l)){
+	  t = (i->length() - (i->end() - (o+l)));
+	  t = t > r ? r : t;
+	  tmp.substr_of(i->data(), 0, t);
+	  ecc_extent tmp_extent(i->offset(), tmp);
+	  out->insert(tmp_extent);
+	  r -=t;
+	  break;
+	}
+	t = i->length();
+	t = t > r ? r : t;
+	tmp.substr_of(i->data(), 0, t);
+	ecc_extent tmp_extent(i->offset(), tmp);
+	out->insert(tmp_extent);
+	r -=t;
+      }
+      return l-r;
+    }
+};
+
+typedef std::set<ecc_object> Cache;
+
+class ECCache
+{
+  private:
+    Mutex lock;
+    uint64_t current_cap;
+    uint64_t cap;
+    uint64_t high_threshold;
+    uint64_t low_threashold;
+    Cache cache;
+
+  public:
+    ECCache(uint64_t c = 104857600, uint64_t ht = 98304000, uint64_t lt = 52428800) : 
+      lock("ECCache::lock"), current_cap(0), cap(c), high_threshold(ht), low_threashold(lt) {
+      cache.clear();
+    }
+
+    ~ECCache(){}
+
+    int set_evict(std::string key)
+    {
+      //Mutex::Locker l(lock);
+      Cache::iterator Cit = cache.find(key);
+      ecc_object TmpCacheObj = *Cit;
+      cache.erase(key);
+      TmpCacheObj.evict();
+      cache.insert(TmpCacheObj);
+      return 0;
+    }
+
+    void trim()
```

Comment:
need is_full() to trigger roll-forward(TBD)
logic to trim objects between high_threshold and low threshold(TBD)
lock wait impl/lock free(TBD)